firmware/lm32/heartbeat.c

```diff
@@ -0,0 +1,77 @@
+
+#include <generated/csr.h>
+#include <generated/mem.h>
+#include <hw/flags.h>
+#include <system.h>
+#include <time.h>
+
+#include "heartbeat.h"
+#include "processor.h"
+#include "hdmi_in0.h"
+#include "hdmi_in1.h"
+#include "pattern.h"
+
+static int heartbeat_status = 0;
+
+void hb_status(int val)
+{
+	if(val==1) { 
+		heartbeat_status = 1; 
+	}
+	else {
+		heartbeat_status = 0;
+	}		
+}
+
+void hb_service(int h_active, int v_active, int source)
+{
+	static int last_event;
+	static int counter;		
+	
+	if (heartbeat_status==1) {		
+		if(elapsed(&last_event, identifier_frequency_read()/5)) {
+			if(counter==1) {
+				hb_fill(h_active, v_active, 0, source);
+				counter = 0;
+			}
+			else {
+				hb_fill(h_active, v_active, 1, source);
+				counter = 1;
+			}
+		}
+	}
+}
+
+void hb_fill(int h_active, int v_active, int n, int source)
+{
+	int addr, i, j;
+	volatile unsigned int *framebuffer = (unsigned int *)(MAIN_RAM_BASE + PATTERN_FRAMEBUFFER_BASE);
```

Comment:
Why do you set this to `MAIN_RAM_BASE + PATTERN_FRAMEBUFFER_BASE` here but used `(MAIN_RAM_BASE + pattern_framebuffer_base())` below?